src/libsystemd/sd-varlink/sd-varlink.c

```diff
@@ -4193,27 +4204,55 @@ _public_ int sd_varlink_error_to_errno(const char *error, sd_json_variant *param
                 { SD_VARLINK_ERROR_EXPECTED_MORE,          -EBADE         },
         };
 
+        int r;
+
         if (!error)
                 return 0;
 
         FOREACH_ELEMENT(t, table)
                 if (streq(error, t->error))
                         return t->value;
 
-        if (streq(error, SD_VARLINK_ERROR_SYSTEM) && parameters) {
-                sd_json_variant *e;
+        /* This following tries to reverse the operation sd_varlink_error_errno() applies to turn errnos into
+         * varlink errors */
+        if (!streq(error, SD_VARLINK_ERROR_SYSTEM))
+                return -EBADR;
 
-                e = sd_json_variant_by_key(parameters, "errno");
-                if (sd_json_variant_is_integer(e)) {
-                        int64_t i;
+        if (!parameters)
+                return -EBADR;
 
-                        i = sd_json_variant_integer(e);
-                        if (i > 0 && i < ERRNO_MAX)
-                                return -i;
-                }
+        /* If an origin is set, check if it's Linux, otherwise don't translate */
+        sd_json_variant *e = sd_json_variant_by_key(parameters, "origin");
+        if (e && (!sd_json_variant_is_string(e) ||
+                  !streq(sd_json_variant_string(e), "linux")))
+                return -EBADR;
+
+        /* If a name is specified, go by name */
+        e = sd_json_variant_by_key(parameters, "name");
+        if (e) {
+                if (!sd_json_variant_is_string(e))
+                        return -EBADR;
+
+                r = errno_from_name(sd_json_variant_string(e));
+                if (r < 0)
+                        return -EBADR;
+
+                assert(r > 0);
+                return -r;
```

Comment:
I'd say that `errno` should still take priority over the textual form for linux. Also, if parsing of the name fails, we should fall back to errno and not fail with -EBADR.

Reply:
in my view textual form of `errno` is for humans who wants to see varlink response by naked eye. So, it's not even necessary to parse it. For other "origins" this may be different though.

Reply:
> I'd say that errno should still take priority over the textual form for linux. Also, if parsing of the name fails, we should fall back to errno and not fail with -EBADR.

well, dunno, in theory the kernel/glibc would have the freedom to one day define mappings between error names and numbers differently for each arch. the kernel currently doesn't do that, thankfully, but it's kinda set up for doing something like that (the names are defined in the "asm" include directory in the kernel), and who knows. hence i figured that if we have the name we better use the name, to be independent of the numbers.

after all the API docs (and the posix specs this is base don) focus on the names, not on the numbers, hence we should focus on them, if we have them.

Reply:
christ, that's awful.